File: common/LoanRepayment/LoanRepayment.py
import math


# Function to calculate loan balance: P(t) = P0 * (1 + r)^t
def calculate_loan_balance(P0, r, t):
    return P0 * math.pow((1 + r), t)


# Function to find the minimum time t such that P(t) < P_target
def find_repayment_time(P0, r, P_target):
    if P0 <= 0 or r <= 0 or P_target <= 0:
        raise ValueError("Initial loan amount, interest rate, and target balance must be positive.")

    t = 0  # Start time from 0
    P_t = P0  # Initial loan balance

    # Iteratively calculate until P(t) is less than the target balance
    while P_t >= P_target:
        t += 1  # Increment time by 1 month
        P_t = calculate_loan_balance(P0, r, t)

    return t  # Return the minimum time t


# Main function to get user input and calculate the repayment time
def main():
    # Get user input
    P0 = float(input("请输入初始贷款金额 P0: "))
    r = float(input("请输入月利率 r (例如 0.01 表示 1%): "))
    P_target = float(input("请输入目标余额 P_target: "))

    # Find the minimum repayment time
    time = find_repayment_time(P0, r, P_target)

    # Output the result
    print(f"贷款余额将小于 {P_target} 在时间 t = {time} 月时发生。")


# Run the main function
if __name__ == "__main__":
    main()

File: common/LoanRepayment/LoanRepayment.java
//根据贷款金额、利率和还款周期，计算满足某个目标偿还金额时所需的时间。
import java.util.Scanner;

public class LoanRepayment {

    // 计算贷款余额 P(t) = P0 * (1 + r)^t
    public static double calculateLoanBalance(double P0, double r, double t) {
        return P0 * Math.pow((1 + r), t);
    }

    // 找到最小的时间 t，使得 P(t) < P_target
    public static double findRepaymentTime(double P0, double r, double P_target) {
        if (P0 <= 0 || r <= 0 || P_target <= 0) {
            throw new IllegalArgumentException("Initial loan amount, interest rate, and target balance must be positive.");
        }

        double t = 0;  // 时间从0开始
        double P_t = P0;  // 初始贷款余额

        // 迭代计算直到 P(t) 小于目标值
        while (P_t >= P_target) {
            t += 1;  // 逐步增加时间
            P_t = calculateLoanBalance(P0, r, t);
        }

        return t;  // 返回最小时间 t
    }

    public static void main(String[] args) {
        // 获取用户输入
        Scanner scanner = new Scanner(System.in);
        System.out.print("请输入初始贷款金额 P0: ");
        double P0 = scanner.nextDouble();

        System.out.print("请输入月利率 r (例如 0.01 表示 1%): ");
        double r = scanner.nextDouble();

        System.out.print("请输入目标余额 P_target: ");
        double P_target = scanner.nextDouble();

        // 找到最小时间 t
        double time = findRepaymentTime(P0, r, P_target);

        // 输出结果
        System.out.println("贷款余额将小于 " + P_target + " 在时间 t = " + time + " 月时发生。");
    }
}
